internal/mongoutil/mongoutil.go

```diff
@@ -0,0 +1,124 @@
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package mongoutil
+
+import (
+	"reflect"
+
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// Options defines options types that can be merged using the functional
+// setters.
+type Options interface {
+	options.AggregateOptions | options.BucketOptions | options.BulkWriteOptions |
+		options.ClientOptions | options.ClientEncryptionOptions | options.CollectionOptions |
+		options.CountOptions | options.CreateIndexesOptions |
+		options.CreateCollectionOptions | options.CreateSearchIndexesOptions |
+		options.CreateViewOptions | options.DataKeyOptions | options.DatabaseOptions |
+		options.DefaultIndexOptions | options.DeleteOptions | options.DistinctOptions |
+		options.DropCollectionOptions | options.DropIndexesOptions |
+		options.DropSearchIndexOptions | options.EncryptOptions |
+		options.EstimatedDocumentCountOptions | options.FindOptions |
+		options.FindOneOptions | options.FindOneAndDeleteOptions |
+		options.FindOneAndReplaceOptions | options.FindOneAndUpdateOptions |
+		options.GridFSFindOptions | options.GridFSNameOptions | options.GridFSUploadOptions |
+		options.IndexOptions | options.InsertManyOptions | options.InsertOneOptions |
+		options.ListCollectionsOptions | options.ListDatabasesOptions |
+		options.ListIndexesOptions | options.ListSearchIndexesOptions |
+		options.LoggerOptions | options.RangeOptions | options.ReplaceOptions |
+		options.RewrapManyDataKeyOptions | options.RunCmdOptions |
+		options.SearchIndexesOptions | options.ServerAPIOptions | options.SessionOptions |
+		options.TimeSeriesOptions | options.TransactionOptions | options.UpdateOptions |
+		options.UpdateSearchIndexOptions | options.ChangeStreamOptions |
+		options.AutoEncryptionOptions
+}
+
+// OptionsBuilder is an interface that wraps a method to return a list of setter
+// functions for merging options for a generic options type.
+type OptionsBuilder[T Options] interface {
+	OptionsSetters() []func(*T) error
+}
+
+// NewOptionsFromBuilder will functionally merge a slice of mongo.Options in a
+// "last-one-wins" manner, where nil options are ignored.
+func NewOptionsFromBuilder[T Options](opts ...OptionsBuilder[T]) (*T, error) {
+	args := new(T)
+	for _, opt := range opts {
+		if opt == nil || reflect.ValueOf(opt).IsNil() {
+			// Do nothing if the option is nil or if opt is nil but implicitly cast as
+			// an Options interface by the NewArgsFromOptions function. The latter
+			// case would look something like this:
+			continue
+		}
+
+		for _, setArgs := range opt.OptionsSetters() {
+			if setArgs == nil {
+				continue
+			}
+
+			if err := setArgs(args); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return args, nil
+}
+
+// OptionsBuilderWithCallback implements a mongo.OptionsBuilder object for an
+// arbitrary options type. The intended use case is to create options from
+// options.
+type OptionsBuilderWithCallback[T Options] struct {
```

Comment:
There is already a constructor for this, `NewBuilderFromOptions`. I can extend it to include callbacks, but the reason we export `OptionsBuilderWithCallback` is because it makes it easier to create a slice of builders with callbacks: https://github.com/prestonvasquez/mongo-go-driver/blob/1248c26ff39d103f56505d6261fb99e7173cacf4/internal/integration/unified/collection_operation_execution.go#L1179-L1208

I can extend `mongoutil.NewBuilderFromOptions` to include a callback option, however.

Reply:
I think we should still export and return the OptionsBuilderWithCallback struct, so as not to break the dependency inversion principle. I will extend the existing `NewBuilderFromOptions` function signature to include a callback argument.